sdk/core/azure-core/inc/nullable.hpp

```diff
@@ -0,0 +1,175 @@
+      T m_value;
+    };
+
+    bool m_hasValue;
+
+    void Destroy() noexcept /* enforces termination */
+    {
+      if (m_hasValue)
+      {
+        m_value.~T();
+        m_hasValue = false;
+      }
+    }
+
+  public:
+    constexpr Nullable() : m_disengaged{}, m_hasValue(false) {}
+    constexpr Nullable(const T& initialValue) : m_value(initialValue), m_hasValue(true) {}
+
+    Nullable(const Nullable& other) noexcept(std::is_nothrow_copy_constructible<T>::value)
+        : m_hasValue(other.m_hasValue)
+    {
+      if (m_hasValue)
+      {
+        // standard implementation has to use std::addressof instead of & :)
+        ::new (static_cast<void*>(&m_value)) T(other.m_value);
+      }
+    }
+
+    Nullable(Nullable&& other) noexcept(std::is_nothrow_move_constructible<T>::value)
+        : m_hasValue(other.m_hasValue)
+    {
+      if (m_hasValue)
+      {
+        // standard implementation has to use std::addressof instead of & :)
+        ::new (static_cast<void*>(&m_value)) T(std::move(other.m_value));
+      }
+    }
+
+    ~Nullable()
+    {
+      if (m_hasValue)
+      {
+        m_value.~T();
+      }
+    }
+
+    void Reset() noexcept /* enforces termination */ { Destroy(); }
+
+    // this assumes that swap can't throw if T is nothrow move constructible because
+    // is_nothrow_swappable is added in C++17
+    void Swap(Nullable& other) noexcept(std::is_nothrow_move_constructible<T>::value)
+    {
+      if (m_hasValue)
+      {
+        if (other.m_hasValue)
+        {
+          using std::swap;
+          swap(m_value, other.m_value);
+        }
+        else
+        {
+          ::new (static_cast<void*>(&other.m_value)) T(std::move(m_value)); // throws
+          other.m_hasValue = true;
+          Destroy();
+        }
+      }
+      else if (other.m_hasValue)
+      {
+        ::new (static_cast<void*>(&m_value)) T(std::move(other.m_value)); // throws
+        m_hasValue = true;
+        other.Destroy();
+      }
+    }
+
+    // Intentionally lowercase to override the swap
+    friend void swap(Nullable& lhs, Nullable& rhs) noexcept(
+        std::is_nothrow_move_constructible<T>::value)
+    {
+      lhs.Swap(rhs);
+    }
+
+    Nullable& operator=(const Nullable& other)
+    {
+      // this copy and swap may be inefficient for some Ts but
+      // it's a lot less code than the standard implementation :)
+      Nullable{other}.Swap(*this);
+      return *this;
+    }
+
+    Nullable& operator=(Nullable&& other) noexcept(std::is_nothrow_move_constructible<T>::value)
+    {
+      // this move and swap may be inefficient for some Ts but
+      // it's a lot less code than the standard implementation :)
+      Nullable{std::move(other)}.Swap(*this);
+      return *this;
+    }
+
+    template <
+        class U = T,
+        typename std::enable_if<
+            !std::is_same<Nullable, std::remove_cv_t<std::remove_reference_t<U>>>::value // Avoid repeated assignment
```

Comment:
This constraint is coped from a forwarding assignment operator but you're not forwarding here. 

All the _ts should not compile in C++14 mode.

Reply:
(I'm extremely nervous about copying one overload of the std::optional overload set without copying every overload; that is a finely tuned 'mess')